### ISSUE TITLE
[SMTChecker] Loop invariants

### ISSUE THREAD

**Issue by leonardoalt** [ethereum/solidity#6210]
It would be nice if the developer is able to give logical hints about loops.
Potential ideas:
- Loop invariants
- Loop bounds
- Loop pre/post conditions

**Comment by wilcoxjay** [comment]
I like the [Dafny](https://github.com/Microsoft/dafny) syntax for loop invariants. Something like this:

```
uint x = 1;
uint y = 0;

for (uint i = 0; i < 100; i++) 
    invariant x % 2 == 1
    invariant y % 2 == 0
{
    x = x + y;
    y = y + 2;
}

// postcondition of loop:
assert x % 2 == 1;
```

**Comment by chriseth** [comment]
@wilcoxjay the location of the invariant sounds good. Would that automatically be an inductive invariant if specified in a loop at that point? Would the postcondition be redundant in that case?

**Comment by wilcoxjay** [comment]
Yes, I usually think of inductive loop invariants. And yes, the postcondition is redundant. I just put it there to clarify that loop postconditions are just asserts.

**Comment by chriseth** [comment]
Considering a loop that has a loop condition which depends on data external to the loop - would the invariant also be checked if the loop does not execute for a single iteration? Phrased differently: Can we assume the loop invariant to be true after the loop even if it did not execute at all?

**Comment by wilcoxjay** [comment]
The way to think about a loop invariant like this is that it should be true every time the loop condition is checked. So in the `for` loop example above, every time `i < 100` is evaluated (no matter whether it comes back true or false), the invariant should be true.

So even if the loop is executed 0 times, the condition is still checked once, so yes, the loop invariant must be true.

**Comment by leonardoalt** [comment]
With Solidity's recent custom Natspec tags, doing something like the following examples annotates the AST node of the loop with the given documentation right before the syntactic loop. Currently loops are not part of dev docs, but imo they could be added to output this information semi-parsed.
```
contract D {
	function f(uint x, uint y) public pure {
		require(x < y);
		/// @custom:formal invariants {x <= y}
		for (uint i = x; i < y; ++i)
			++x;
		assert(x == y);
	}
}
```

**Comment by leonardoalt** [comment]
Since this is/could be used by various tools without changing the language itself, maybe this is an easier and more practical way to go.
